Add Type Inference for $_ / $PSItem in catch{ } blocks

vexx32 opened the pull request:
## PR Summary

Fix #7828 

* Adds type inference results to `$_` / `$PSItem` when it is used to refer to the `ErrorRecord` in a catch block.
* Adds test for the type inference.
* Includes type inference for typed `catch [ExceptionType] { $_ }` blocks such that `$_` is inferred as a dummy `ErrorRecord<TException>` which is identical to `ErrorRecord` but with a `TException`-typed `Exception` member.
  * In other words, using typed catch blocks correctly autocompletes the members for `$_.Exception.<tab>`

Example:
```powershell
PS> try {throw 'ball'} catch {$_.<tab>
# completes to:
PS> try {throw 'ball'} catch {$_.CategoryInfo
```

(Compare current behaviour, where tab completion is missing due to a lack of type inference on the `$_` / `$PSItem` variable.)

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
Fix tab completions

vexx32 committed:
Add tests for type inference of $_ in catch blocks

iSazonov commented:
Perhaps we need a test that `try {throw 'ball'} catch {$d=Get-Date; $_.CategoryInfo` works.

lzybkr reviewed (approved):
It would be cool to also add special handling for:
```powershell
try {}
catch [SomeSpecificException] { $_.Exception.<TAB> }
```
I realize this could be a bit messy though and I'm not sure it's commonly needed, so maybe it's not worth it.

vexx32 commented:
Can you expand upon that? As-is, the currently available logic handles this as you'd expect, completing the members for a generic exception object.

Are there some exceptions with additional members we would want to try to detect from the exception type handed to the catch block itself?

I don't think this would go quite in the same place on this code, so if you could point out where it would need to be added, I'll give it a whirl.

iSazonov commented:
@vexx32 Please add test to demo that both cases work.

vexx32 committed:
Add test for System.Exception completion from un-typed catch block

Add dummy generic errorrecord class
for type inference

Add logic for inferring errorrecord/exception type

Add tests for typed catch block inference

Tidy Exception<E> class

vexx32 commented:
@lzybkr @iSazonov 

Added tests.

With @SeeminglyScience's help, *also* added in some type inference for typed catch blocks. He came up with the idea of a dummy generic `ErrorRecord<T>` class that could be used just for type inference in order to get the members of the typed exceptions. 

Pretty happy with it! 😄

vexx32 committed:
Reformat constructor for StyleCop

Update generic type argument
Update tests.
(thanks @SeeminglyScience!)

lzybkr commented:
Clever but simple, perfect. 👍👍👍

iSazonov commented on test/powershell/engine/Api/TypeInference.Tests.ps1:
```diff
@@ -926,6 +926,33 @@ Describe "Type inference Tests" -tags "CI" {
+        $variableAst = { try {} catch { $_ } }.Ast.Find({ param($a) $a -is [System.Management.Automation.Language.VariableExpressionAst] }, $true)
+        $res = [AstTypeInference]::InferTypeOf($variableAst)
+
+        $res.Count | Should -Be 1
```
`$res | Should -HaveCount 1`

iSazonov replied:
Could you please address the comment?
Line 941 too.

vexx32 replied:
All done. There are a few tests for other things that should probably use this as well, but I think that might be better left to another PR specifically for test revisions.

iSazonov commented on test/powershell/engine/Api/TypeInference.Tests.ps1:
```diff
@@ -926,6 +926,33 @@ Describe "Type inference Tests" -tags "CI" {
+    }
+
+    It 'Infers type of $_.Expression in typed catch block' {
+        $memberAst = { try {} catch [System.IO.FileNotFoundException] { $_.Exception } }.Ast.Find(
```
Please add more catch-es.

vexx32 replied:
Reformatted this for additional Test Cases with a variety of exception types.

Would you like me to add additional tests verifying that constructs like `catch [type1], [type2] {}` and subsequent catch blocks all work as expected?

vexx32 committed:
Add testcases to verify that type inference works with common exceptions

vexx32 committed:
Reformat slightly for less confusing syntax

vexx32 committed:
Add further tests just to be absolutely sure

iSazonov commented on test/powershell/engine/Api/TypeInference.Tests.ps1:
```diff
@@ -942,15 +942,44 @@ Describe "Type inference Tests" -tags "CI" {
+    It 'Infers type of $_.Expression in [<Type>] typed catch block' -TestCases $catchClauseTypes {
+        param([Type] $Type)
+
+        $memberAst = [scriptblock]::Create("try {} catch [$Type] { $_.Exception }").Ast.Find(
```
I meant:
```c#
"try {} catch [System.OverflowException] catch [System.TimeoutException] { $_.Exception }"
```

vexx32 replied:
See additional below tests. I suppose I only test `catch [type] {$_.Exception} catch {$_.Exception}` but if you want me to go through a couple stacks I'm game. :)

vexx32 committed:
Fix tests
Reformat test strings
Remove [type] specifier from params (not needed)

vexx32 commented:
Appveyor caught something funky, don't think it was related. Rebased to restart it.

Other than that... everything looks like it works as it ought. 😄

vexx32 committed:
Fix tests to use -HaveCount
